### ISSUE TITLE
CRASH (8.0.18523) dr_insert_clean_call with small immediate operand

### ISSUE THREAD

**Issue by clslgrnc** [DynamoRIO/dynamorio#4452]
**Describe the bug**
The [documentation of `dr_insert_clean_call`](https://dynamorio.org/dynamorio_docs/dr__ir__utils_8h.html#ae7b7bd1e750b8a24ebf401fb6a6d6d5e) currently states that:
> The routine also supports immediate integers that are smaller than the register size

Yet on x64 a `dr_insert_clean_call` with an operand of size 8 or 16 leads to a crash.

**To Reproduce**
From the root of the repository, at the currently last commit on branch master 6c638197ce9e651812f33b4f31f7710d94f0cc6d (tag: cronbuild-8.0.18523).

Modify [suite/tests/client-interface/count-bbs.dll.c](https://github.com/DynamoRIO/dynamorio/blob/6c638197ce9e651812f33b4f31f7710d94f0cc6d/suite/tests/client-interface/count-bbs.dll.c) to test `dr_insert_clean_call` with a 16 bits operand:

```diff
--- a/suite/tests/client-interface/count-bbs.dll.c
+++ b/suite/tests/client-interface/count-bbs.dll.c
@@ -38,7 +38,7 @@ static uint64 bbcnt1 = 0;
 static uint64 bbcnt4 = 0;
 
 static void
-bbcount4(reg_t r1, reg_t r2, reg_t r3, reg_t r4)
+bbcount4(reg_t r1, reg_t r2, reg_t r3, u_int16_t r4)
 {
     bbcnt4++;
 }
@@ -74,7 +74,7 @@ bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace, bool trans
                          opnd_create_reg(REG_XAX));
     dr_insert_clean_call(drcontext, bb, instr, bbcount4, false, 4,
                          opnd_create_reg(REG_XAX), opnd_create_reg(REG_XBX),
-                         opnd_create_reg(REG_XCX), opnd_create_reg(REG_XDX));
+                         opnd_create_reg(REG_XCX), OPND_CREATE_INT16(0));
     return DR_EMIT_DEFAULT;
 }
```
Build and test:
```console
$ cmake -DBUILD_TESTS=ON ../dynamorio
$ make -j6
$ ctest -V -R "count-bbs"
test 127
    Start 127: code_api|client.count-bbs

127: Test command: /dynamorio/bin64/drrun "-s" "90" "-quiet" "-killpg" "-stderr_mask" "0xC" "-dumpcore_mask" "0" "-code_api" "-c" "/dynamorio/suite/tests/bin/libclient.count-bbs.dll.so" "--" "/dynamorio/suite/tests/bin/simple_app"
127: Test timeout computed to be: 1500
127: <Application /dynamorio/suite/tests/bin/simple_app (259721).  Tool internal crash at PC 0x00007f1d70a6be1d.  Please report this at your tool's issue tracker.  Program aborted.
127: Received SIGSEGV at generated pc 0x00007f1d70a6be1d in thread 259721
127: Base: 0x00007f1db4a5a000
127: Registers:eax=0x0000000000000000 ebx=0x00007f1b70b05720 ecx=0x0000000000000000 edx=0x0000000000000000
127: 	esi=0x0000000000000000 edi=0x00007f1b70b05700 esp=0x00007f1b70b05778 ebp=0x0000000000000000
127: 	r8 =0x0000000000000000 r9 =0x0000000000000000 r10=0x0000000000000000 r11=0x0000000000000000
127: 	r12=0x0000000000000000 r13=0x0000000000000000 r14=0x0000000000000000 r15=0x0000000000000000
127: 	eflags=0x0000000000010202
127: version 8.0.18523, custom build
127: -no_dynamic_options -client_lib '/dynamorio/suite/tests/bin/libclient.count-bbs.dll.so;0;' -client_lib64 '/dynamorio/suite/tests/bin/libclient.count-bbs.dll.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_sy
127: 0x00007f1b70adf080 0x0000000000000000>
1/1 Test #127: code_api|client.count-bbs ........***Failed  Required regular expression not found. Regex=[^Hello, world!
bbcount0_fp\.\.\. yes
bbcount1\.\.\. yes
bbcount4\.\.\. yes
$
]  0.03 sec

0% tests passed, 1 tests failed out of 1
```

**Expected behavior**
The modified test should pass, or the doc should be modified to state more explicitly what operands are not supported.

**Versions**
 - What version of DynamoRIO are you using? **8.0.1**
 - Does the latest build from
https://github.com/DynamoRIO/dynamorio/wiki/Latest-Build solve the problem? **No**
- What operating system version are you running on? **Ubuntu 20.04.1 LTS**
- Is your application 32-bit or 64-bit? **64**

**Additional context**
Might be a duplicate of #1707.

**Comment by derekbruening** [comment]
Thank you for the detailed filing.

The intent is to support smaller sizes, so I would say it's a bug in the code rather than the docs.

It is surprising that it hits SIGSEGV instead of an encoder error or sthg.  I wonder what the details are there.

If the example code is the actual desired code, using an immediate, there is a simple workaround of changing the immediate size of course.

If you would like to submit a pull request to fix this that would be most appreciated.

**Comment by clslgrnc** [comment]
I am probably not familiar enough with DynamoRIO's internals to fix this myself.

TL;DR DynamoRIO's machine code for `mov    $0x0000000000000000 -> %rcx` with a 16 bits operand is wrong.

--- 

Running the example with `-debug -loglevel 3` we get:
```
computing memory target for 0x00007f.......edd causing SIGSEGV, kernel claims it is 0x0000000000000000
For SIGSEGV at cache pc 0x00007f.......edd, computed target read 0x0000000000000000
    faulting instr: vmovdqa 0x40(%rsp)[32byte] -> %ymm0
```

Looking at registers: `xsp=0x00007f.......778`. `RSP` is not aligned on 32 bytes and `vmovdqa` raises a GPF.

I think this comes from a wrong assembly:
```
after instrumentation:
TAG  0x00007f.......100
 +0    m4 @0x00007f.......f68  65 48 a3 00 00 00 00 mov    %rax -> %gs:0x00[8byte]
                               00 00 00 00
 +11   m4 @0x00007f.......b20  65 48 a1 20 00 00 00 mov    %gs:0x20[8byte] -> %rax
                               00 00 00 00
 +22   m4 @0x00007f.......ba0  48 89 60 18          mov    %rsp -> 0x18(%rax)[8byte]
 +26   m4 @0x00007f.......580  48 8b a0 48 09 00 00 mov    0x00000948(%rax)[8byte] -> %rsp
 +33   m4 @0x00007f.......4b0  65 48 a1 00 00 00 00 mov    %gs:0x00[8byte] -> %rax
                               00 00 00 00
 +44   m4 @0x00007f.......600  48 8d a4 24 f8 f6 ff lea    0xfffff6f8(%rsp) -> %rsp
                               ff
 +52   m4 @0x00007f.......9a0  e8 bb 74 f6 ff       call   $0x00007f.......dc0 %rsp -> %rsp 0xfffffff8(%rsp)[8byte]
 +57   m4 @0x00007f.......e00  48 8b 44 24 38       mov    0x38(%rsp)[8byte] -> %rax
 +62   m4 @0x00007f.......c20                       <label>
 +62   m4 @0x00007f.......8b8  66 48 c7 c1 00 00    mov    $0x0000000000000000 -> %rcx
 +68   m4 @0x00007f.......c20  48 8b 54 24 30       mov    0x30(%rsp)[8byte] -> %rdx
 +73   m4 @0x00007f.......aa0  48 8b f3             mov    %rbx -> %rsi
 +76   m4 @0x00007f.......348  48 8b f8             mov    %rax -> %rdi
 +79   m4 @0x00007f.......6d0  e8 d4 68 f5 ff       call   $0x00007f.......1d9 %rsp -> %rsp 0xfffffff8(%rsp)[8byte]
 +84   m4 @0x00007f.......ca0  e8 bb 75 f6 ff       call   $0x00007f.......ec0 %rsp -> %rsp 0xfffffff8(%rsp)[8byte]
 +89   m4 @0x00007f.......e68  65 48 a3 00 00 00 00 mov    %rax -> %gs:0x00[8byte]
                               00 00 00 00
 +100  m4 @0x00007f.......750  65 48 a1 20 00 00 00 mov    %gs:0x20[8byte] -> %rax
                               00 00 00 00
 +111  m4 @0x00007f.......ee8  48 8b 60 18          mov    0x18(%rax)[8byte] -> %rsp
 +115  m4 @0x00007f.......050  65 48 a1 00 00 00 00 mov    %gs:0x00[8byte] -> %rax
                               00 00 00 00
 +126  m4 @0x00007f.......880                       <label>
 +126  L3 @0x00007f.......ca0  48 89 e7             mov    %rsp -> %rdi
 +129  L3 @0x00007f.......920  e8 e8 0c 00 00       call   $0x00007f.......df0 %rsp -> %rsp 0xfffffff8(%rsp)[8byte]
END 0x00007f.......100
```

Looking more precisely to the lines that insert the parameters:
```
 +62   m4 @0x00007f.......8b8  66 48 c7 c1 00 00    mov    $0x0000000000000000 -> %rcx
 +68   m4 @0x00007f.......c20  48 8b 54 24 30       mov    0x30(%rsp)[8byte] -> %rdx
 +73   m4 @0x00007f.......aa0  48 8b f3             mov    %rbx -> %rsi
 +76   m4 @0x00007f.......348  48 8b f8             mov    %rax -> %rdi
```
If I try to disassemble the machine code with https://defuse.ca/online-x86-assembler.htm, I get:
```
0:  66 48 c7 c1 00 00 48    data16 mov rcx,0xffffffff8b480000
7:  8b
8:  54                      push   rsp
9:  24 30                   and    al,0x30
b:  48 8b f3                mov    rsi,rbx
e:  48 8b f8                mov    rdi,rax 
```
The spurious `push rsp` would explain the misalignment. I guess `defuse.ca` is right and DynamoRIO's machine code for `mov    $0x0000000000000000 -> %rcx` with a 16 bits operand is wrong.

**Comment by clslgrnc** [comment]
As a side note, `dr_insert_clean_call_ex` does not seem to honor `DR_CLEANCALL_NOSAVE_XMM` (or even `DR_CLEANCALL_NOSAVE_XMM | DR_CLEANCALL_NOSAVE_XMM_NONPARAM | DR_CLEANCALL_NOSAVE_XMM_NONRET`): `YMM` registers are still saved and restored with this flag.

**Comment by clslgrnc** [comment]
`dr_insert_clean_call_ex` tries to insert a `data16 mov rcx,0x0000` as:
```
66 48 c7 c1 00 00
```

I do not think there is a `mov r64, imm16`, instead DynamoRIO should sign extend the 16bits operand and generate a `mov r64, imm32`:
```
66 48 c7 c1 00 00 00 00
```

Or better yet a `mov r16, imm16` since the callee only uses the 16 lower bits.
